## Review by Kai

## PR Review: feat(memory) write-side mode-switch verification (#435)

After four prior review rounds, the code is in good shape. Two minor new findings below.

### Critical

None.

### Warning

None.

### Suggestion

**`test_bot.py` - `return_exceptions=True` hides task failure in the positive test**

In `test_extraction_invoked_when_memory_enabled`:

```python
if new_tasks:
    await asyncio.gather(*new_tasks, return_exceptions=True)
extract_mock.assert_called_once()
```

If `_ingest_memory` raises before reaching `extract_and_store` (e.g., a config guard or missing attribute raises mid-task), the exception is swallowed by `return_exceptions=True` and the failure surfaces only as `AssertionError: Expected 'mock' to have been called once` with no indication of the underlying cause. In the disabled-mode counterpart, swallowing a task exception is safe because a task appearing at all is already the regression; in the enabled-mode test, a crashing task and a missing task are indistinguishable from the assertion message. Removing `return_exceptions=True` here propagates the task exception directly, making the failure unambiguous. The disabled-mode test is fine as-is.

---

**`test_eval_modeswitch.py` - `_attach_parents` docstring contradicts the type-ignore it accompanies**

```python
child.parent = parent  # type: ignore[attr-defined]
```

The docstring says "pyright tolerates it under the strict-but-permissive attribute model," but the `# type: ignore[attr-defined]` on the same line proves the opposite: pyright flags it as an error and the ignore is silencing that error. The comment is wrong and would mislead a reader deciding whether the ignore is load-bearing. Suggest replacing "pyright tolerates it" with "pyright flags this as an attr-defined error on strict AST nodes, hence the ignore."

---

### Previously raised, not re-raised

The `kai.memory.is_enabled` vs `kai.bot.memory_is_enabled` monkeypatch target mismatch in the two `TestHandleResponse` tests was raised in the first review pass. Code is unchanged; not re-raised.

### Overall

The `_capture_recall_logs` context manager is correctly structured: fresh handler and list per invocation, `prior_propagate` saved and restored unconditionally in `finally`, `removeHandler` in `finally`. The `_parse_recall_reason` `continue`-on-ValueError fix from the prior round is present. The deferred `from kai.memory import _RECALL_REASON_DISABLED` is placed after the prior `_isolated_data_dir` calls in execution order, so `kai.memory` is already in `sys.modules` when it runs. `test_callback_disabled_short_circuits` uses the correct monkeypatch target (`memory_command.memory.is_enabled`) for the module-qualified call form in `memory_command.py`. No critical or warning-level new findings.

--- a/src/kai/eval/modeswitch.py
+++ b/src/kai/eval/modeswitch.py
@@ -505,15 +505,17 @@ async def _run_verify() -> int:
             # on (a missing or wrongly-tagged record under disabled
             # mode would silently break those harnesses).
             #
-            # Deferred import for the same reason `_isolated_data_dir`
-            # imports `kai.memory` lazily: a top-level import would
-            # capture `kai.memory` in `sys.modules` BEFORE the
-            # MEM0_DIR redirect at the start of `_run_verify` could
-            # take effect, deadlocking the harness against the live
-            # `~/.mem0/migrations_qdrant` lock the running service
-            # holds. The `if "kai.memory" in sys.modules:` guard at
-            # the top of `_run_verify` enforces the invariant; this
-            # deferred form preserves it.
+            # Deferred to preserve the `if "kai.memory" in sys.modules:`
+            # invariant guarded at the top of `_run_verify`. A
+            # top-level `from kai.memory import _RECALL_REASON_DISABLED`
+            # in this module would put `kai.memory` into `sys.modules`
+            # before `_run_verify` runs, tripping the assertion. The
+            # constant itself is a string literal and does not boot
+            # Mem0 or touch the migrations_qdrant lock; the deadlock
+            # concern that motivates `_isolated_data_dir`'s lazy
+            # import does not apply directly to this constant. The
+            # consistency-with-the-guard rationale is what carries
+            # the deferral.
             from kai.memory import _RECALL_REASON_DISABLED
 
             with _isolated_data_dir(tmp_root):

--- a/tests/test_bot.py
+++ b/tests/test_bot.py
@@ -2916,7 +2916,19 @@ class TestHandleResponse:
         # extract_and_store call. The OUTER is_enabled() guard is the
         # contract under test; the inner check would mask its failure
         # if both were False.
-        config = _make_config(memory_extraction_enabled=True)
+        #
+        # episode_classifier_context_turns=0 mirrors the enabled-mode
+        # counterpart's setup so the two tests are structurally
+        # symmetric. Under current behavior the disabled test never
+        # reaches the inner block, so the value is unread; the
+        # symmetry is defense-in-depth: if the monkeypatch ever stops
+        # taking effect or the guard moves, the inner block would
+        # otherwise execute with the default value of 3 and trigger
+        # an unstated `get_recent_pairs(12345, 4)` disk read.
+        config = _make_config(
+            memory_extraction_enabled=True,
+            episode_classifier_context_turns=0,
+        )
         ctx = _make_context(config=config, pool=pool)
 
         # Snapshot the task set BEFORE the call so we can compute the

--- a/tests/test_bot.py
+++ b/tests/test_bot.py
@@ -2992,9 +2992,18 @@ class TestHandleResponse:
         # the coroutine onto the event loop; the test's coroutine
         # cannot assert until the scheduled task has had a chance to
         # progress past its `await extract_and_store(...)` line.
+        #
+        # NOT `return_exceptions=True`: in the positive test, a task
+        # that crashed BEFORE reaching extract_and_store (e.g., a
+        # config-guard mismatch, a missing attribute) would be
+        # silently swallowed and surface only as "expected called
+        # once, called zero times" - hiding the real cause. The
+        # disabled-mode counterpart uses return_exceptions=True
+        # because there a task appearing at all is the regression
+        # signal; here we want crashes to propagate.
         new_tasks = asyncio.all_tasks() - tasks_before - {asyncio.current_task()}
         if new_tasks:
-            await asyncio.gather(*new_tasks, return_exceptions=True)
+            await asyncio.gather(*new_tasks)
 
         extract_mock.assert_called_once()
 

--- a/tests/test_eval_modeswitch.py
+++ b/tests/test_eval_modeswitch.py
@@ -376,11 +376,11 @@ class TestExtractionCallSiteGating:
     def _attach_parents(tree: ast.AST) -> None:
         """Attach a `parent` attribute to every node so ancestor
         traversal works after `ast.walk`. ast.AST does not declare
-        a `parent` field, so the assignment is silently typed as
-        `Any`; pyright tolerates it under the strict-but-permissive
-        attribute model. This is the standard pattern; the
-        alternative (a side-table dict keyed by id(node)) is more
-        verbose with no semantic gain.
+        a `parent` field, so pyright flags the assignment as an
+        attr-defined error; the `# type: ignore[attr-defined]`
+        below is load-bearing, not cosmetic. The alternative
+        (a side-table dict keyed by id(node)) is more verbose with
+        no semantic gain; this is the standard ast-walking pattern.
         """
         for parent in ast.walk(tree):
             for child in ast.iter_child_nodes(parent):

--- a/tests/test_eval_modeswitch.py
+++ b/tests/test_eval_modeswitch.py
@@ -300,6 +300,16 @@ class TestRecallReasonField:
         # boot cost (~80MB embedding model on first run); the
         # production path under verification is format_context, not
         # init_memory.
+        #
+        # `object()` is safe ONLY because format_context never
+        # invokes attributes on `_memory` directly: `is_enabled()`
+        # checks `_memory is not None`, and every search call
+        # routes through the module-level `search()` wrapper, which
+        # is mocked below. A future refactor that called something
+        # like `_memory.client_ready()` from the hot path would
+        # break this test with a confusing `AttributeError` rather
+        # than a clean assertion failure; the sentinel would need
+        # to grow into a proper Mem0 mock at that point.
         memory_module._memory = object()
         memory_module._config = modeswitch._build_test_configs(memory_enabled_value=True)
 

